README.md
### kitForm

kitForm is an extension for the Customer Relationship Management (CRM) [KeepInTouch] [3] and the Content Management Systems [WebsiteBaker] [1] or [LEPTON CMS] [2]. It enables the easy creation of **forms** and **dialogs** for the CMS and the CRM. 

#### Requirements

* minimum PHP 5.2.x
* using [WebsiteBaker] [1] 2.8.x _or_ [LEPTON CMS] [2] 1.x
* [dbConnect_LE] [3] installed 
* [Dwoo] [4] installed
* [DropletsExtension] [5] installed
* [KeepInTouch] [6] installed
* [kitTools] [7] installed
* [wbLib] [8] installed
* jQueryAdmin **uninstalled** (_deprecated_, replaced by [LibraryAdmin] [9])
* [LibraryAdmin] [9] installed
* [libJQuery] [9] installed

#### Downloads

* [dbConnect_LE] [10]
* [Dwoo] [11]
* [DropletsExtension] [12]
* [KeepInTouch] [13]
* [kitTools] [14]
* [wbLib] [15]
* [LibraryAdmin] [16]
* [libJQuery] [17]
* [kitForm] [18]

#### Installation

* download the actual [kitForm] [18] installation archive
* in CMS backend select the file from "Add-ons" -> "Modules" -> "Install module"

#### First Steps

In the backend select "Admin-Tools" -> "kitForm". You will find around a dozen predefined sample dialogs and forms with brief descriptions. Use this samples as base for your own forms and dialogs.

To use a form in frontend please remember the name of the form, i.e. `kit_feedback` and edit the page where to place the form with your WYSIWYG editor.

`kit_feedback` can be used to enable the visitor of your website to give you feedback, it shows a form to send feedback, shows the feedback thread, gather feedbacks at the KIT event log a.s.o.

**kitForm** uses the Droplet `kit_form` to place the forms, type in your WYSIWYG editor:

    [[kit_form?form=kit_feedback]]
    
that's really all, the form is placed at your website and will work!  

Please visit the [phpManufaktur] [19] to get more informations about **kitForm** and join the [Addons Support Group] [20].

[1]: http://websitebaker2.org "WebsiteBaker Content Management System"
[2]: http://lepton-cms.org "LEPTON CMS"
[3]: https://addons.phpmanufaktur.de/dbConnect_LE
[4]: https://addons.phpmanufaktur.de/Dwoo
[5]: https://addons.phpmanufaktur.de/DropletsExtension
[6]: https://addons.phpmanufaktur.de/KeepInTouch
[7]: https://addons.phpmanufaktur.de/kitTools
[8]: http://wblib.webbird.de/de/
[9]: http://jquery.lepton-cms.org/
[10]: https://addons.phpmanufaktur.de/download.php?file=dbConnect_LE
[11]: https://addons.phpmanufaktur.de/download.php?file=Dwoo
[12]: https://addons.phpmanufaktur.de/download.php?file=DropletsExtension
[13]: https://addons.phpmanufaktur.de/download.php?file=KeepInTouch
[14]: https://addons.phpmanufaktur.de/download.php?file=DropletsExtension
[15]: https://github.com/webbird/wblib/downloads
[16]: http://jquery.lepton-cms.org/modules/download_gallery/dlc.php?file=75&id=1318585713
[17]: http://jquery.lepton-cms.org/modules/download_gallery/dlc.php?file=76&id=1320743410
[18]: https://addons.phpmanufaktur.de/download.php?file=kitForm
[19]: https://addons.phpmanufaktur.de/kitForm
[20]: https://phpmanufaktur.de/support


CHANGELOG
### kitForm CHANGELOG

Ralf Hertsch <ralf.hertsch@phpmanufaktur.de>
http://phpmanufaktur.de
(c) 2011-2013 - phpManufaktur by Ralf Hertsch

Release 0.44 - 2014-05-21

* fixed: A SQL injection error within the sorter.php file can be exploited by
  unauthenticated people to conduct SQL injection attacks.

Release 0.43 - 2013-11-22

* fixed missing parts: problem accessing language files at BlackCat CMS environment
* added icon.png (BlackCat CMS) and tool_icon.png (WebsiteBaker 2.8.4)

Release 0.42 - 2013-11-21

* fixed: the CSV export does not support the combined field kit_zip_city
* fixed: problem accessing language files at BlackCat CMS environment
* changed: precheck.php does no longer check for Dwoo and DropletsExtension at
  BlackCat CMS environment

Release 0.41 - 2013-01-03

* added: admin function to remove duplicate submitted forms
* added: delete form data record in protocol detail view
* added: use protocol_limit in the config.json to change the item limit in the
  protocol list, see also <http://yals.de/OQRv0Y> for more informations
* added: admin function to delete unpublished feedbacks (SPAM) and optional
  also the associated KIT ID's
* fixed: the internal precheck for KIT may fail because it does not use the
  custom table prefix if defined
* added: admin function to export submitted form data as CSV file

Release 0.40 - 2012-12-08

* fixed: free fields and notes are mapped in the wrong way by kitForm
* added: added field names in the response array of the forms
* fixed: standard file uploads are added twice to the protocol
* added: support for special upload links from KeepInTouch 0.64

Release 0.39 - 2012-11-16

* fixed: the kit_form droplet does not check if the TABLE_PREFIX has changed
* added: kit_form droplet can now redirect the user after a successful login
  (http://yals.de/SM25dY)
* fixed: if missing an KIT ID in the kitForm protocol it will now prompt the
  missing ID but still continue with the protocol
* fixed: check if key exists before access value (class.backend.php)
* added: wait the specified seconds before the form submission is accepted
* fixed: duplicate entries and processing of emails due calls of DropletsExtension
* changed usage of the KIT contact title: the form gets now the value defined in
  the KIT list settings. Please use now:
  {if $contact.kit_title == 'Herr'}r{/if} {$contact.kit_title}
  instead of
  {if $contact.kit_title == 'titleMister'}r Herr{else} Frau{/if}

Release 0.38 - 2012-08-04

* the prefix "custom." enables custom templates for class.backend.php
* fixed: too strong field validation in class.frontend.php, please read
  <http://yals.de/QoSbQ5> for more informations
* fixed a problem reading markdown formatted texts in feedback.htt

Release 0.37 - 2012-07-22

* added ability to change the table prefix easy via config.json, please read
  <http://yals.de/OQRv0Y> for more informations
* moved template plugin use_markdown() to Dwoo
* moved template plugin translate() to Dwoo

Release 0.36 - 2012-07-13

* changed initializing of the Dwoo template engine
* added a Dwoo plugin for formatting Markdown text
* added jQuery plugin for character count and limit in textareas
* removed redirecting to about dialog from upgrade.php
* added a Dwoo plugin for I18n support

Release 0.35 - 2012-06-14

* removed WebsiteBaker version check

Release 0.34 - 2012-06-07

* moved repository from SVN to GitHub
* added CHANGELOG
* changed license from GPL to MIT
* added LICENSE
* removed info.txt
* added README.md
* changed all header and archive informations
* fixed: the kit_feedback form may prompt an error while sending status mails if
  an user account is inactive or deleted - now still continue.
* fixed: using parse_str() in class.backend.php for additional fields may
  ignore &amp;
* fixed a redirection problem after successfull installation of kitForm

Release 0.33 - 2012-02-29

* added: additional information which newsletter was registered for the forms
  mail.client.register.newsletter.htt and mail.provider.register.newsletter.htt
  please read /kit_form/htt/changes.txt for more informations
* added: kitForm checks the KIT option cfgNewsletterAccountInfo

Release 0.32 - 2012-02-19

* added: new KIT datafield "Contact language" enables to automatically set the
  contact language of the form sender
* changed: standard template form.htt, please read /kit_form/htt/changes.txt for
  more informations
* changed: introduce $_SESSION['KIT_HTML_REQUEST'] to mark fields which should
  not processed by xssPrevent()

Release 0.31 - 2012-02-16

* added: new standard dialog kit_password_change enables the user to change the
  password
* added: new KIT LINK to force password changes if the system detect insecure
  passwords

Release 0.30 - 2012-02-04

* added: KIT support for country selection
* added: KIT support for birthday date
* added: fallback for presets and languages: it is no longer neccessary that all
  forms for all languages really exists, kitForm now falls back to the standard
  preset in /kit_form/htt/1 in three steps.
* added: switches in the Droplet [[kit_form]] for language, fallback_language,
  fallback_preset and debug (see also Droplet hints)
* removed the default preset in /kit_form/htt/2 because it is to much work to
  keep this nearly with preset 1 identical preset up-to-date
* added: debug switch in all templates, please read /kit_form/htt/changes.txt
  for more informations

Release 0.29 - 2012-02-03

* added: entries for fields of type 'select', 'checkbox' and 'radio' can now
  be moved and change their position.

Release 0.28 - 2012-02-03

* added: check of option 'relaying' for email providers, handle emails in a
  different way and use the reply-to header

Release 0.27 - 2011-12-19

* fixed: precheck.php looks if kitDirList is installed and check the release
  number and kitDirList needs kitForm: so we got an endless loop if kitDirList
  must be updated ...

Release 0.26 - 2011-12-17


precheck.php
<?php

/**
 * kitForm
 *
 * @author Ralf Hertsch <ralf.hertsch@phpmanufaktur.de>
 * @link http://phpmanufaktur.de
 * @copyright 2011 - 2012
 * @license MIT License (MIT) http://www.opensource.org/licenses/MIT
 */

// include class.secure.php to protect this file and the whole CMS!
if (defined('WB_PATH')) {
  if (defined('LEPTON_VERSION'))
    include(WB_PATH.'/framework/class.secure.php');
}
else {
  $oneback = "../";
  $root = $oneback;
  $level = 1;
  while (($level < 10) && (!file_exists($root.'/framework/class.secure.php'))) {
    $root .= $oneback;
    $level += 1;
  }
  if (file_exists($root.'/framework/class.secure.php')) {
    include($root.'/framework/class.secure.php');
  }
  else {
    trigger_error(sprintf("[ <b>%s</b> ] Can't include class.secure.php!", $_SERVER['SCRIPT_NAME']), E_USER_ERROR);
  }
}
// end include class.secure.php

// Checking Requirements
global $database;

$checked = true;

// check PHP version
$PRECHECK['PHP_VERSION'] = array(
    'VERSION' => '5.2.0',
    'OPERATOR' => '>='
    );

// check dependencies for KIT at runtime but not at installation!
$PRECHECK['KIT'] = array(
    'kit' => array('VERSION' => '0.61', 'OPERATOR' => '>='),
    'kit_dirlist' => array('VERSION' => '0.28', 'OPERATOR' => '>=')
);

// modified precheck array
$check = array(
    'dbConnect_LE' => array(
        'directory' => 'dbconnect_le',
        'version' => '0.68',
        'problem' => 'dbConnect_LE => <b><a href="https://addons.phpmanufaktur.de/download.php?file=dbConnect_LE" target="_blank">Download actual version</a></b>'
        ),
    'KeepInTouch' => array(
        'directory' => 'kit',
        'version' => '0.10',
        'problem' => 'KeepInTouch => <b><a href="https://addons.phpmanufaktur.de/download.php?file=KeepInTouch" target="_blank">Download actual version</a></b>'
        ),
    'kitTools' => array(
        'directory' => 'kit_tools',
        'version' => '0.16',
        'problem' => 'kitTools => <b><a href="https://addons.phpmanufaktur.de/download.php?file=kitTools" target="_blank">Download actual version</a></b>'
        ),
    'wbLib' => array(
        'directory' => 'wblib',
        'version' => '0.65',
        'problem' => 'wbLib => <b><a href="https://github.com/webbird/wblib/downloads" target="_blank">Download actual version</a></b>'
        ),
    'LibraryAdmin' => array(
        'directory' => 'libraryadmin',
        'version' => '1.9',
        'problem' => 'LibraryAdmin => <b><a href="http://jquery.lepton-cms.org/modules/download_gallery/dlc.php?file=75&id=1318585713" target="_blank">Download actual version</a></b>'
        ),
    'libJQuery' => array(
        'directory' => 'lib_jquery',
        'version' => '1.25',
        'problem' => 'libJQuery => <b><a href="http://jquery.lepton-cms.org/modules/download_gallery/dlc.php?file=76&id=1320743410" target="_blank">Download actual version</a></b>'
        ),
    );

if (!defined('CAT_VERSION')) {
    $check['Dwoo'] = array(
        'directory' => 'dwoo',
        'version' => '0.16',
        'problem' => 'Dwoo => <b><a href="https://addons.phpmanufaktur.de/download.php?file=Dwoo" target="_blank">Download actual version</a></b>'
    );
    $check['dropletsExtension'] = array(
        'directory' => 'droplets_extension',
        'version' => '0.24',
        'problem' => 'dropletsExtension => <b><a href="https://addons.phpmanufaktur.de/download.php?file=dropletsExtension" target="_blank">Download actual version</a></b>'
    );
}
$versionSQL = "SELECT `version` FROM `".TABLE_PREFIX."addons` WHERE `directory`='%s'";

foreach ($check as $name => $addon) {
  // loop throug the addons and check the versions
  $version = $database->get_one(sprintf($versionSQL, $addon['directory']), MYSQL_ASSOC);
  if (false === ($status = version_compare(!empty($version) ? $version : '0', $addon['version'], '>='))) {
    $checked = false;
    $key = $addon['problem'];
  }
  else
    $key = $name;
  $PRECHECK['CUSTOM_CHECKS'][$key] = array(
      'REQUIRED' => $addon['version'],
      'ACTUAL' => !empty($version) ? $version : '- not installed -',
      'STATUS' => $status
      );
}

// check default charset
$SQL = "SELECT `value` FROM `".TABLE_PREFIX."settings` WHERE `name`='default_charset'";
$charset = $database->get_one($SQL, MYSQL_ASSOC);
if ($charset != 'utf-8') {
  $checked = false;
  $key = 'This addon needs UTF-8 as default charset!';
}
else
  $key = 'UTF-8';

$PRECHECK['CUSTOM_CHECKS'][$key] = array(
    'REQUIRED' => 'utf-8',
    'ACTUAL' => $charset,
    'STATUS' => ($charset == 'utf-8')
    );

// jQueryAdmin should be uninstalled
if (file_exists(WB_PATH . '/modules/jqueryadmin/tool.php')) {
  $checked = false;
  $key = 'jQueryAdmin is <b>deprecated</b>, please uninstall and<br />use <b>LibraryAdmin</b> instead!';
}
else
  $key = 'jQueryAdmin';

$PRECHECK['CUSTOM_CHECKS'][$key] = array(
    'REQUIRED' => 'REMOVED',
    'ACTUAL' => (file_exists(WB_PATH . '/modules/jqueryadmin/tool.php')) ? 'INSTALLED' : 'REMOVED',
    'STATUS' => (!file_exists(WB_PATH . '/modules/jqueryadmin/tool.php'))
    );

if (!$checked) {
  // if a problem occured prompt a hint and grant that the LEPTON/WB precheck fail
  $PRECHECK['CUSTOM_CHECKS']['Please install or update all required addons.<br />Need help? Please contact the <b><a href="https://phpmanufaktur.de/support" target="_blank">phpManufaktur Support Group</a></b>.'] = array(
      'REQUIRED' => 'OK',
      'ACTUAL' => 'PROBLEM',
      'STATUS' => false
      );
}
